src/test/serfloat_tests.cpp

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2014-2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <test/util/setup_common.h>
+#include <util/serfloat.h>
+
+#include <boost/test/unit_test.hpp>
+
+#include <cmath>
+#include <limits>
+
+BOOST_FIXTURE_TEST_SUITE(serfloat_tests, BasicTestingSetup)
+
+namespace {
+
+uint64_t TestDouble(double f) {
+    uint64_t i = EncodeDouble(f);
+    double f2 = DecodeDouble(i);
+    if (std::isnan(f)) {
+        // NaN is not guaranteed to round-trip exactly.
+        BOOST_CHECK(std::isnan(f2));
+    } else {
+        // Everything else is.
+        BOOST_CHECK(!std::isnan(f2));
+        uint64_t i2 = EncodeDouble(f2);
+        BOOST_CHECK_EQUAL(f, f2);
+        BOOST_CHECK_EQUAL(i, i2);
```

Comment:
would it make sense to also assert equality with the "legacy" serialization helpers in both this unit test and the fuzz test?

Otherwise we might run into a corrupted fees.dat after an upgrade. A problem worse than the problem this pull is trying to solve (making fees.dat architecture independent).

Reply:
Oh sorry, I see that is what the "should produce exactly the same as the in-memory representation for non-NaN" test is doing.

Reply:
Though, it looks like the fuzz test isn't fuzzing the in-memory representation, only the round-trip?

Reply:
Added back.

Reply:
Ah thanks, though the `fuzzed_data_provider.ConsumeFloatingPoint` only consumes "nice" floats, not odd ones. So the fuzzer is only testing the happy path.

Reply:
Fixed in #22180